cocos/platform/CCFileUtils.h

```diff
@@ -41,6 +42,76 @@ NS_CC_BEGIN
  * @{
  */
 
+
+class ResizableBuffer {
+public:
+    virtual ~ResizableBuffer() {}
+    virtual void resize(size_t size) = 0;
+    virtual void* buffer() const = 0;
+};
+
+template<typename T>
+class ResizableBufferAdapter { };
+
+
+template<typename C, typename T, typename A>
+class ResizableBufferAdapter< std::basic_string<C, T, A> > : public ResizableBuffer {
+    typedef std::basic_string<C, T, A> BufferType;
+    BufferType* buffer_;
+public:
+    ResizableBufferAdapter(BufferType* buffer) : buffer_(buffer) {}
+    virtual void resize(size_t size) override {
+        buffer_->resize((size + sizeof(C)) / sizeof(C));
+    }
+    virtual void* buffer() const override {
+        return &buffer_->front();
+    }
+};
+
+template<typename T, typename A>
+class ResizableBufferAdapter< std::vector<T, A> > : public ResizableBuffer {
+    typedef std::vector<T, A> BufferType;
+    BufferType* buffer_;
+public:
+    ResizableBufferAdapter(BufferType* buffer) : buffer_(buffer) {}
+    virtual void resize(size_t size) override {
+        buffer_->resize((size + sizeof(T)) / sizeof(T));
+    }
+    virtual void* buffer() const override {
+        return &buffer_->front();
+    }
+};
+
+
+template<>
+class ResizableBufferAdapter<Data> : public ResizableBuffer {
+    typedef Data BufferType;
+    BufferType* buffer_;
+public:
+    ResizableBufferAdapter(BufferType* buffer) : buffer_(buffer) {}
+    virtual void resize(size_t size) override {
+        if (buffer_->getSize() < size) {
+            auto old = buffer_->getBytes();
+            void* buffer = realloc(old, size);
+            if (buffer)
+                buffer_->fastSet((unsigned char*)buffer, size);
+        }
+    }
+    virtual void* buffer() const override {
+        return buffer_->getBytes();
+    }
+};
+
+enum class FileError
+{
+    OK = 0,
+    NotExists = 0x01, /** File not exists */
+    OpenFailed = 0x02, /** Open file failed. */
+    ReadFaild = 0x03, /** Read failed */
+    NotInitialized = 0x04, /** FileUtils is not initializes */
+};
```

Comment:
Should be OK.
(actually I was initially wrote the error in FileUtils, but move out for shorten the return type).